src/mapport.cpp

```diff
@@ -32,13 +41,117 @@ static_assert(MINIUPNPC_API_VERSION >= 10, "miniUPnPc API version >= 10 assumed"
+    LogPrintf("natpmp: initnatpmp() failed with %d error.\n", r_init);
+    return false;
+}
+
+static bool NatpmpDiscover(natpmp_t* natpmp, struct in_addr& external_ipv4_addr)
+{
+    const int r_send = sendpublicaddressrequest(natpmp);
+    if (r_send == 2 /* OK */) {
+        int r_read;
+        natpmpresp_t response;
+        do {
+            r_read = readnatpmpresponseorretry(natpmp, &response);
+        } while (r_read == NATPMP_TRYAGAIN);
+
+        if (r_read == 0) {
+            external_ipv4_addr = response.pnu.publicaddress.addr;
+            return true;
+        } else if (r_read == NATPMP_ERR_NOGATEWAYSUPPORT) {
+            LogPrintf("natpmp: The gateway does not support NAT-PMP.\n");
+        } else {
+            LogPrintf("natpmp: readnatpmpresponseorretry() for public address failed with %d error.\n", r_read);
+        }
+
+    } else {
+        LogPrintf("natpmp: sendpublicaddressrequest() failed with %d error.\n", r_send);
+    }
+
+    return false;
+}
+
+static bool NatpmpMapping(natpmp_t* natpmp, const struct in_addr& external_ipv4_addr, uint16_t private_port, bool& external_ip_discovered)
+{
+    const uint16_t suggested_external_port = g_mapport_external_port ? g_mapport_external_port : private_port;
+    const int r_send = sendnewportmappingrequest(natpmp, NATPMP_PROTOCOL_TCP, private_port, suggested_external_port, 3600 /*seconds*/);
+    if (r_send == 12 /* OK */) {
+        int r_read;
+        natpmpresp_t response;
+        do {
+            r_read = readnatpmpresponseorretry(natpmp, &response);
+        } while (r_read == NATPMP_TRYAGAIN);
+
+        if (r_read == 0) {
+            auto pm = response.pnu.newportmapping;
+            if (private_port == pm.privateport && pm.lifetime > 0) {
+                g_mapport_external_port = pm.mappedpublicport;
+                const CService external{external_ipv4_addr, pm.mappedpublicport};
+                if (!external_ip_discovered && fDiscover) {
+                    AddLocal(external, LOCAL_MAPPED);
+                    external_ip_discovered = true;
+                }
+
+                LogPrintf("natpmp: Port mapping successful. External address = %s\n", external.ToString());
+                return true;
+            } else {
+                LogPrintf("natpmp: Port mapping failed.\n");
+            }
+
+        } else if (r_read == NATPMP_ERR_NOGATEWAYSUPPORT) {
+            LogPrintf("natpmp: The gateway does not support NAT-PMP.\n");
+        } else {
+            LogPrintf("natpmp: readnatpmpresponseorretry() for port mapping failed with %d error.\n", r_read);
+        }
+
+    } else {
+        LogPrintf("natpmp: sendnewportmappingrequest() failed with %d error.\n", r_send);
+    }
+
+    return false;
+}
+
+static bool ProcessNatpmp()
+{
+    bool ret = false;
+    natpmp_t natpmp;
+    struct in_addr external_ipv4_addr;
+    if (NatpmpInit(&natpmp) && NatpmpDiscover(&natpmp, external_ipv4_addr)) {
+        bool external_ip_discovered = false;
```

Comment:
https://github.com/bitcoin/bitcoin/pull/18077/commits/f6bd717f3f44f2a167ce2e548df1132789d0ea98

Hm, I see we're explicitly setting this in `NetpmpMapping` but then it goes unused. No big deal, just curious.

Reply:
As an in-out parameter, `external_ip_discovered` is used here https://github.com/bitcoin/bitcoin/blob/81a791d88d68e64968e52d8ff6ec20879bd5fad1/src/mapport.cpp#L136-L138 to prevent consecutive `AddLocal` calls.

Reply:
Oh I see, thanks.